editor/plugins/script_text_editor.cpp

```diff
@@ -1362,11 +1362,38 @@ void ScriptTextEditor::_edit_option(int p_op) {
 			code_editor->remove_all_bookmarks();
 		} break;
 		case DEBUG_TOGGLE_BREAKPOINT: {
-			for (int caret_idx = 0; caret_idx < tx->get_caret_count(); caret_idx++) {
-				int line = tx->get_caret_line(caret_idx);
-				bool dobreak = !tx->is_line_breakpointed(line);
-				tx->set_line_as_breakpoint(line, dobreak);
-				EditorDebuggerNode::get_singleton()->set_breakpoint(script->get_path(), line + 1, dobreak);
+			Vector<int> caret_edit_order = tx->get_caret_index_edit_order();
+			caret_edit_order.reverse();
+			int last_line = -1;
+			for (const int &c : caret_edit_order) {
+				int from = tx->has_selection(c) ? tx->get_selection_from_line(c) : tx->get_caret_line(c);
+				from += from == last_line ? 1 : 0;
+				int to = tx->has_selection(c) ? tx->get_selection_to_line(c) : tx->get_caret_line(c);
+				if (to < from) {
+					continue;
+				}
+				// Check first if there's any lines with breakpoints in the selection.
+				bool selection_has_breakpoints = false;
+				for (int line = from; line <= to; line++) {
+					if (tx->is_line_breakpointed(line)) {
+						selection_has_breakpoints = true;
+						break;
+					}
+				}
+
+				// Set breakpoint on caret or remove all bookmarks from the selection.
+				if (!selection_has_breakpoints) {
+					if (tx->get_caret_line(c) != last_line) {
+						tx->set_line_as_breakpoint(tx->get_caret_line(c), true);
+						EditorDebuggerNode::get_singleton()->set_breakpoint(script->get_path(), tx->get_caret_line(c) + 1, true);
```

Comment:
I don't think we need to call `EditorDebuggerNode::get_singleton()->set_breakpoint` from here anymore. Should now be caught by `void ScriptTextEditor::_breakpoint_toggled(int p_row)`

Reply:
Seems to be working without it.